encodedv/encodedv.1
.\" encodedv - encode a series of images to a digital video stream
.\" Copyright (c) 2001 James Bowman, Peter Schlaile, Daniel Kobras
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page was written especially for Debian Linux. It is based
.\" on encodedv's help output.
.\"
.TH ENCODE 1 "April 2001"
.SH NAME
encodedv \- encode a series of images to a digital video stream
.SH SYNOPSIS
.B encodedv
[ \fBoptions\fR ] \fIvideo_pattern\fR \fI[ audio_input ]\fR
.SH DESCRIPTION
\fBencodedv\fR takes a series of images in PPM/PGM/video format and
optionally an audio stream and compiles them
into a DV-encoded video stream. \fIpattern\fR specifies where the input data is
located and may be either \fI-\fR for stdin, or a \fBprintf\fR-style pattern
taking one numeric argument that gets replaced by the current frame number.
Like image%03d.ppm, which gets expanded to image000.ppm, image001.ppm, etc.
.TP
\fB\-\-version\fR
show encodedv version number.
.TP
\fB\-s\fR, \fB\-\-start\-frame\fR=\fIcount\fR
start encoding at frame number \fIcount\fR (defaults to 0).
.TP
\fB\-e\fR, \fB\-\-end\-frame\fR=\fIcount\fR
end encoding at frame number \fIcount\fR (defaults to unlimited).
.TP
\fB\-l\fR, \fB\-\-wrong\-interlace\fR
flip lines to compensate for wrong interlacing in the input data. Happened
with mpeg2dec generated data.
.TP
\fB\-p\fR, \fB\-\-vlc\-passes\fR
vlc code distribution passes (1-3) greater values = better quality but not 
necessarily slower encoding! This defaults for best quality = 3.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
show encoder statistics / status information
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIfilter-name\fR
Choose input-filter: [>ppm<, pgm, video]
The \fBppm\fR-filter only supports raw rgb ppm files. 
The \fBpgm\fR file format is the one generated by \fBmpeg2dec\fR of the
livid project. ( http://linuxvideo.org ) This means: Y-data appended by
U and V data which are scaled down by 2 and placed side by side. This
option defaults to PPM. Some things you want to keep in mind:

1) If you want to stream
video frames (using video_pattern = "-") you have to make sure that there 
is no trailing garbage at the end of the pictures. This is ignored by most 
image manipulation programs!

2) If you are one of the poor persons with a buggy PCI bus-mastering board
be sure to have a backup handy if you want to use video support. Some versions
of the VIA board chipsets crash your machine and even your harddisk. Since
encodedv uses full resolution capturing it is more likely to trigger these
bugs than other programs. If you have a VIA board you definitely want to
upgrade to Linux 2.4.3 or higher.

3) The encoded pictures must have the correct resolution. If they don't, you
may want to try ppmqscale. The used video format is chosen by picture 
resolution. Use 720x576 for PAL and 720x480 for NTSC.

4) If you want to speed up things a little bit and you are generating the
input pictures automatically you may want to try pgm's instead of ppm's since
they are encoded somewhat faster. But keep in mind, that this pgm format is
only optimal for PAL since NTSC averages the U and V values somewhat 
differently.
.TP
\fB\-a\fR, \fB\-\-audio\-input\fR=\fIfilter-name\fR
Choose audio-input-filter: [>none<, wav, dsp]
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIfilter-name\fR
Choose output-filter: [>raw<]
Hint for Arne Schirmacher to add AVI support ;-)
.TP
\fB\-q\fR, \fB\-\-static\-qno\fR=\fItable-no\fR
Static qno tables for quantisation on 2 VLC passes. 
For turbo (but somewhat lossy encoding) try -q [1,2] -p [2,3].
There are only two static qno tables registered right now:

1 : for sharp DV pictures

2 : for somewhat noisy satelite television signal

If you want to add some more, go ahead ;-)
.TP
\fB\-f\fR, \fB\-\-fps\fR=\fIfps-number\fR
Set frames per second (default: use all frames)
.TP
\fB\-d, \fB\-\-force\-dct\fR=\fIdct-mode\fR
Force dct mode (88 or 248) for whole picture
.PP
Help Options
.TP
-\fB?\fR, \fB\-\-help\fR
Show help message.
.TP
\fB\-\-usage\fR
Display brief usage message.
.SH "SEE ALSO"
.BR playdv (1).
.SH AVAILABILITY
See http://libdv.sourceforge.net/ for the latest version.
.SH AUTHORS
.B encodedv
was written by James Bowman <jamesb@excamera.com> and
Peter Schlaile <udbz@rz.uni-karlsruhe.de>.
.PP
This manual page is based on encodedv's help message and was written by
Daniel Kobras <kobras@debian.org> for the Debian GNU/Linux system
(but may be used by others). It was updated 
by Peter Schlaile <udbz@rz.uni-karlsruhe.de>.



NEWS
The best place to keep up with events is to check the libdv mailing
lists:

	http://sourceforge.net/mail/?group_id=4393

NOTE: as of the time of writing, 1/14/2001, the libdv-dev archive is
not working.  I'm trying to resolve it with sourceforge, but it is
frustratingly hard to get a response from them.


encodedv/encodedv.c
/*
 *  encodedv.c
 *
 *     Copyright (C) James Bowman  - July 2000
 *                   Peter Schlaile - Jan 2001
 *
 *  This file is part of libdv, a free DV (IEC 61834/SMPTE 314M)
 *  codec.
 *
 *  libdv is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU Lesser Public License as published by
 *  the Free Software Foundation; either version 2.1, or (at your
 *  option) any later version.
 *   
 *  libdv is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser Public License for more details.
 *   
 *  You should have received a copy of the GNU Lesser Public License
 *  along with libdv; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. 
 *
 *  The libdv homepage is http://libdv.sourceforge.net/.  
 */

#include <string.h>
#include "libdv/dv_types.h"
#include "libdv/dv.h"
#include "libdv/encode.h"
#include "libdv/enc_input.h"
#include "libdv/enc_output.h"
#include "libdv/headers.h"

#define DV_ENCODER_OPT_VERSION         0
#define DV_ENCODER_OPT_START_FRAME     1
#define DV_ENCODER_OPT_END_FRAME       2
#define DV_ENCODER_OPT_WRONG_INTERLACE 3
#define DV_ENCODER_OPT_ENCODE_PASSES   4
#define DV_ENCODER_OPT_VERBOSE         5
#define DV_ENCODER_OPT_INPUT           6
#define DV_ENCODER_OPT_AUDIO_INPUT     7
#define DV_ENCODER_OPT_OUTPUT          8
#define DV_ENCODER_OPT_AUTOHELP        9
#define DV_ENCODER_OPT_STATIC_QNO      10
#define DV_ENCODER_OPT_FPS             11
#define DV_ENCODER_OPT_FORCE_DCT       12
#define DV_ENCODER_OPT_16X9            13
#define DV_ENCODER_OPT_STDIN           14
#define DV_ENCODER_NUM_OPTS            15

int main(int argc, char *argv[])
{
	unsigned long start = 0;
	unsigned long end = 0xfffffff;
	int wrong_interlace = 0;
	const char* filename = NULL;
	const char* audio_filename = NULL;
	int vlc_encode_passes = 3;
	int static_qno = 0;
	int verbose_mode = 0;
	const char* audio_input_filter_str = "none";
	const char* input_filter_str = "ppm";
	const char* output_filter_str = "raw";
	dv_enc_audio_input_filter_t * audio_input_filter = NULL;
	dv_enc_input_filter_t * input_filter;
	dv_enc_output_filter_t * output_filter;
	int count = 0;
	int fps = 0;
	int is16x9 = 0;
	int err_code;
	int force_dct = -1;
	int	isStdin = 0;

#if HAVE_LIBPOPT
	struct poptOption option_table[DV_ENCODER_NUM_OPTS+1]; 
	int rc;             /* return code from popt */
	poptContext optCon; /* context for parsing command-line options */
	option_table[DV_ENCODER_OPT_VERSION] = (struct poptOption) {
		longName: "version", 
		val: 'v', 
		descrip: "show encode version number"
	}; /* version */

	option_table[DV_ENCODER_OPT_START_FRAME] = (struct poptOption) {
		longName:   "start-frame", 
		shortName:  's', 
		argInfo:    POPT_ARG_INT, 
		arg:        &start,
		argDescrip: "count",
		descrip:    "start at <count> frame (defaults to 0)"
	}; /* start-frame */

	option_table[DV_ENCODER_OPT_END_FRAME] = (struct poptOption) {
		longName:   "end-frame", 
		shortName:  'e', 
		argInfo:    POPT_ARG_INT, 
		arg:        &end,
		argDescrip: "count",
		descrip:    "end at <count> frame (defaults to unlimited)"
	}; /* end-frames */

	option_table[DV_ENCODER_OPT_WRONG_INTERLACE] = (struct poptOption) {
		longName:   "wrong-interlace", 
		shortName:  'l', 
		arg:        &wrong_interlace,
		descrip:    "flip lines to compensate for wrong interlacing"
	}; /* wrong-interlace */

	option_table[DV_ENCODER_OPT_ENCODE_PASSES] = (struct poptOption) {
		longName:   "vlc-passes", 
		shortName:  'p', 
		argInfo:    POPT_ARG_INT, 
		arg:        &vlc_encode_passes,
		descrip:    "vlc code distribution passes (1-3) "
		"greater values = better quality but "
		"not necessarily slower encoding!"
	}; /* vlc encoder passes */

	option_table[DV_ENCODER_OPT_VERBOSE] = (struct poptOption) {
		longName:   "verbose", 
		shortName:  'v', 
		arg:        &verbose_mode,
		descrip:    "show encoder statistics / status information"
	}; /* verbose mode */

	option_table[DV_ENCODER_OPT_INPUT] = (struct poptOption) {
		longName:   "input", 
		shortName:  'i', 
		arg:        &input_filter_str,
		argInfo:    POPT_ARG_STRING, 
		descrip:    "choose input-filter [>ppm<, pgm, video]"
	}; /* input */

	option_table[DV_ENCODER_OPT_AUDIO_INPUT] = (struct poptOption) {
		longName:   "audio-input", 
		shortName:  'a', 
		arg:        &audio_input_filter_str,
		argInfo:    POPT_ARG_STRING, 
		descrip:    "choose audio-input-filter [>none<, wav, dsp]"
	}; /* audio input */

	option_table[DV_ENCODER_OPT_OUTPUT] = (struct poptOption) {
		longName:   "output", 
		shortName:  'o', 
		arg:        &output_filter_str,
		argInfo:    POPT_ARG_STRING, 
		descrip:    "choose output-filter [>raw<]"
	}; /* output */

	option_table[DV_ENCODER_OPT_STATIC_QNO] = (struct poptOption) {
		longName:   "static-qno", 
		shortName:  'q', 
		arg:        &static_qno,
		argInfo:    POPT_ARG_INT, 
		descrip:    "static qno tables for quantisation on 2 VLC "
		"passes. For turbo (but somewhat lossy encoding) try "
		"-q [1,2] -p 2"
	}; /* start qno */

	option_table[DV_ENCODER_OPT_FPS] = (struct poptOption) {
		longName:   "fps", 
		shortName:  'f', 
		arg:        &fps,
		argInfo:    POPT_ARG_INT, 
		descrip:    "set frames per second (default: use all frames)"
	}; /* fps */

	option_table[DV_ENCODER_OPT_FORCE_DCT] = (struct poptOption) {
		longName:   "force-dct", 
		shortName:  'd', 
		arg:        &force_dct,
		argInfo:    POPT_ARG_INT, 
		descrip:    "force dct mode (88 or 248) for whole picture"
	}; /* force dct */

	option_table[DV_ENCODER_OPT_16X9] = (struct poptOption) {
		longName:   "wide", 
		shortName:  'w', 
		arg:        &is16x9,
		argInfo:    POPT_ARG_INT, 
		descrip:    "set wide 16 x 9 format"
	}; /* 16x9 */

	option_table[DV_ENCODER_OPT_STDIN] = (struct poptOption) {
		longName:   "", 
		shortName:  '-', 
		arg:        &isStdin,
		descrip:    "set stdin input"
	}; /* stdin */

	option_table[DV_ENCODER_OPT_AUTOHELP] = (struct poptOption) {
		argInfo: POPT_ARG_INCLUDE_TABLE,
		arg:     poptHelpOptions,
		descrip: "Help options",
	}; /* autohelp */

	option_table[DV_ENCODER_NUM_OPTS] = (struct poptOption) { 
		NULL, 0, 0, NULL, 0 };

	optCon = poptGetContext(NULL, argc, 
				(const char **)argv, option_table, 0);
	poptSetOtherOptionHelp(optCon, "<filename pattern or - for stdin>"
		" [<audio input>]");

	while ((rc = poptGetNextOpt(optCon)) > 0) {
		switch (rc) {
		case 'v':
			fprintf(stderr,"encode: version %s, "
				"http://libdv.sourceforge.net/\n",
				"CVS 01/14/2001");
			exit(0);
			break;
		default:
			break;
		} /* switch */
	} /* while */

	if (rc < -1) {
		/* an error occurred during option processing */
		fprintf(stderr, "%s: %s\n",
			poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
			poptStrerror(rc));
		exit(-1);
	}

	if ( !isStdin ) {
		filename = poptGetArg(optCon);
		if(filename == NULL) {
			poptPrintUsage(optCon, stderr, 0);
			fprintf(stderr, 
				"\nSpecify a single <filename pattern> argument; "
				"e.g. pond%%05d.ppm or - for stdin\n"
				"(For audio input specify additionally "
				"the audio source.)\n");
			exit(-1);
		}
	}
	else {
		filename = "-";
	}
 
	audio_filename = poptGetArg(optCon);
	poptFreeContext(optCon);

#else
	/* No popt, no usage and no options!  HINT: get popt if you don't
	 * have it yet, it's at: ftp://ftp.redhat.com/pub/redhat/code/popt 
	 */
	filename = argv[1];
#endif /* HAVE_LIBPOPT */

	dv_enc_get_input_filters(&input_filter, &count);
	while (count 
	       && strcmp(input_filter->filter_name, input_filter_str) != 0) {
		input_filter++;
		count--;
	}
	if (!count) {
		fprintf(stderr, "Unknown input filter selected: %s!\n"
			"The following filters are supported:\n",
			input_filter_str);
		dv_enc_get_input_filters(&input_filter, &count);
		while (count--) {
			fprintf(stderr, "%s\n", input_filter->filter_name);
			input_filter++;
		}
		return(-1);
	}

	if (strcmp(audio_input_filter_str, "none") != 0) {
		dv_enc_get_audio_input_filters(&audio_input_filter, &count);
		while (count 
		       && strcmp(audio_input_filter->filter_name, 
				 audio_input_filter_str) != 0) {
			audio_input_filter++;
			count--;
		}
		if (!count) {
			fprintf(stderr, 
				"Unknown audio input filter selected: %s!\n"
				"The following filters are supported:\n",
				audio_input_filter_str);
			dv_enc_get_audio_input_filters(&audio_input_filter, 
						       &count);
			while (count--) {
				fprintf(stderr, "%s\n", 
					audio_input_filter->filter_name);
				audio_input_filter++;
			}
			return(-1);
		}
		if (!audio_filename) {
			fprintf(stderr, "Audio input selected but no filename "
				"or device given!\n");
			return(-1);
		}
	}

	dv_enc_get_output_filters(&output_filter, &count);
	while (count && 
	       strcmp(output_filter->filter_name, output_filter_str) != 0) {
		output_filter++;
		count--;
	}
	if (!count) {
		fprintf(stderr, "Unknown output filter selected: %s!\n"
			"The following filters are supported:\n",
			output_filter_str);
		dv_enc_get_output_filters(&output_filter, &count);
		while (count--) {
			fprintf(stderr, "%s\n", output_filter->filter_name);
			output_filter++;
		}
		return(-1);
	}
	if (static_qno < 0 || static_qno > 2) {
		fprintf(stderr, "There are only two static qno tables "
			"registered right now:\n"
			"1 : for sharp DV pictures\n"
			"2 : for somewhat noisy satelite television signal\n"
			"\nIf you want to add some more, go ahead ;-)\n");
		return(-1);
	}
	if (static_qno && vlc_encode_passes == 1) {
		fprintf(stderr, "Warning! static_qno enabled but "
			"vlc_passes == 1.\n"
			"This won't gain you anything for now !!!\n");
	}
	if (force_dct != -1) {
		switch (force_dct) {
		case 88:
			force_dct = DV_DCT_88;
			break;
		case 248:
			force_dct = DV_DCT_248;
			break;
		default:
			fprintf(stderr, "--force-dct has to be 88 or 248!\n");
			return(-1);
		}
	}

        dv_init(NULL, FALSE, FALSE);

	if (input_filter->init(wrong_interlace, force_dct) < 0) {
		return -1;
	}
	if (output_filter->init(stdout) < 0) {
		return -1;
	}
	/* audio_input_filter->init() is in dv_encoder_loop! */

	err_code = dv_encoder_loop(input_filter,audio_input_filter, output_filter,
				start, end, filename, audio_filename,
				vlc_encode_passes, static_qno,
				verbose_mode, fps, is16x9);

	input_filter->finish();
	if (audio_input_filter) {
		audio_input_filter->finish();
	}
	output_filter->finish();
  
	if (verbose_mode) {
		dv_show_statistics();
	}
	return 0;
}


